### ISSUE TITLE
Modules with same names as built in types causes ICE

### ISSUE THREAD

**Issue by matthewjasper** [Rust-GCC/gccrs#3315]
### Code

```rust
mod i32 {}
```

### Meta

Version: gccrs (Compiler-Explorer-Build-gcc-b5c354d038f800695a8d730c56c4a4f744134adb-binutils-2.42) 14.0.1 20240309 (experimental)

### Error output

```
crab1: internal compiler error: in setup_builtin_types, at rust/resolve/rust-late-name-resolver-2.0.cc:97
0x27eb5ec internal_error(char const*, ...)
	???:0
0xaf6cc9 fancy_abort(char const*, int, char const*)
	???:0
0xdbb522 Rust::Resolver2_0::Late::go(Rust::AST::Crate&)
	???:0
0xc4bef0 Rust::Session::compile_crate(char const*)
	???:0
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
Compiler returned: 1
```

**Comment by liamnaddell** [comment]
From reading here: https://doc.rust-lang.org/reference/names/namespaces.html

builtins and modules both belong in the Types namespace, however, that can't really be correct, since `rustc` allows you to name a module `i32`. I guess the idea is that builtins are resolved separately from other types? Otherwise, how would you prevent NR conflicts here?

**Comment by liamnaddell** [comment]
As a note, `rustc` does NOT let you write the following:

```rust
liam@newgame ~/projects/gccrs/gccrs-build $ cat main.rs
pub mod i32 {
    pub fn a() {}
}

struct i32 {}

fn main() {
    i32::a();
}
liam@newgame ~/projects/gccrs/gccrs-build $ rustc main.rs
error[E0428]: the name `i32` is defined multiple times
 --> main.rs:5:1
  |
1 | pub mod i32 {
  | ----------- previous definition of the module `i32` here
...
5 | struct i32 {}
  | ^^^^^^^^^^ `i32` redefined here
  |
  = note: `i32` must be defined only once in the type namespace of this module
```

**Comment by liamnaddell** [comment]
I guess what we are expected to do is remove builtin types from the resolution context when another `Item` overshadows it

**Comment by liamnaddell** [comment]
Modules are clearly handled separately from other types like structs. I think modules are somehow part of their own namespace?

**Comment by powerboat9** [comment]
It looks like builtins are really supposed to go in a [language prelude](https://doc.rust-lang.org/reference/names/preludes.html#language-prelude). That way, module definitions can shadow builtin definitions by virtue of being in a different scope.

**Comment by liamnaddell** [comment]
> It looks like builtins are really supposed to go in a [language prelude](https://doc.rust-lang.org/reference/names/preludes.html#language-prelude). That way, module definitions can shadow builtin definitions by virtue of being in a different scope.

Would this be part of a distinct RIB?

**Comment by powerboat9** [comment]
Yep. Fixing this would probably entail some adjustments to the `ForeverStack` node hierarchy, and some code for handling decent into prelude nodes (since we'll want multiple preludes).

**Comment by liamnaddell** [comment]
> Yep. Fixing this would probably entail some adjustments to the `ForeverStack` node hierarchy, and some code for handling descent into prelude nodes (since we'll want multiple preludes).

Wouldn't the prelude end up being handled as an import? I.e. not having a distinct node?

https://github.com/rust-lang/rust/blob/8ad2c9724d983cfb116baab0bb800edd17f31644/compiler/rustc_resolve/src/imports.rs#L75
https://github.com/rust-lang/rust/blob/8ad2c9724d983cfb116baab0bb800edd17f31644/compiler/rustc_resolve/src/lib.rs#L1021

**Comment by powerboat9** [comment]
There are multiple preludes, some of which will have to have their own nodes

**Comment by liamnaddell** [comment]
I'm giving this one a try currently. Solving the issue seems easy enough, since you just replace insert with insert_globbed. 

Some of the secondary problems are more challenging though.

1.  You aren't able to `use` builtin types, since `self::i32` isn't supposed to exist (theres a rust spec reference for this, I forgot where though). 
2. `struct`s and other proper types shadow builtin types for type resolution
3. `mod`ules DON'T shadow types for type resolution

I'm trying to see currently how many of these I can solve presently

**Comment by powerboat9** [comment]
Modules should be inserted into the type namespace, which would solve 4. We might just need a separate rib/node to store builtin types, and some adjustments to path resolution.

**Comment by liamnaddell** [comment]
This should be fixed now?
